### PR TITLE
Create an interface for handlers

### LINKED ISSUE #31: Handler interface?
I was just wondering why the handlers don't use a common interface? That way if a library uses Guzzle and want to give the option to use an alternative handler, they can just accept for example a RingHandlerInterface object. This has come as somewhat of a requirement due to clients now being immutable. I want to create a Guzzle client with some internal options, but also give the user an option to specify a different handler. What's the best way to do this if an interface doesn't exist?

### PR DESCRIPTION
**Pull request by rtheunissen**
This pull request aims to resolve #31

### CODE REVIEW

File: src/Client/HandlerInterface.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+interface HandlerInterface
+{
+	/**
+     * @param array $request
+     *
+     * @return CompletedFutureArray
+     */
+	public function __invoke(array $request);
+}
```
Comment (GrahamCampbell) [inline src/Client/HandlerInterface.php]: cs

File: src/Client/HandlerInterface.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?php
+namespace GuzzleHttp\Ring\Client;
+
+interface HandlerInterface
+{
+	/**
```
Comment (GrahamCampbell) [inline src/Client/HandlerInterface.php]: cs

**Comment by GrahamCampbell** [comment]
This is technically breaking though

File: src/Client/HandlerInterface.php (hunk 1 above)
Comment (rtheunissen) [inline src/Client/HandlerInterface.php]: Fixed in 23ef734f3f13b8b8d553a15a7a14e5329b7aa9d9

File: src/Client/HandlerInterface.php (hunk 2 above)
Comment (rtheunissen) [inline src/Client/HandlerInterface.php]: Fixed in 23ef734f3f13b8b8d553a15a7a14e5329b7aa9d9

**Comment by rtheunissen** [comment]
What do you by technically breaking?

**Comment by GrahamCampbell** [comment]
"They have to implement HandlerInterfaces"

**Comment by rtheunissen** [comment]
Yeah it's breaking so would need to be a part of a stable release in the future, right?

**Comment by rtheunissen** [comment]
How does that break bc?

**Comment by rtheunissen** [comment]
If it breaks backwards compatibility it would have to be part of the next release.

**Comment by GrahamCampbell** [comment]
I doubt there will be another major version given that guzzle 6.0 isn't too far away? I don't really know though. ;)

**Comment by rtheunissen** [comment]
All current tests pass and it wouldn't affect currently deployed instances as the handlers would still be callable. Couldn't this fall under a minor version?

**Comment by GrahamCampbell** [comment]
Tecnically no, because it does need a semver major version, though I suspect @mtdowling would just do a minor release if this were to be merged, because I don't see anyone actually being affected to badly by this?

**Comment by rtheunissen** [comment]
> MAJOR version when you make incompatible API changes

The API hasn't changed, the behaviour hasn't changed either. Any instance that was type hinting for `callable` would still work. This would give new projects the opportunity to type hint for HandlerInterface, and request that the passed handler be an updated dependancy.

**Comment by rtheunissen** [comment]
Also why does RingPHP have to follow Guzzle's major version?

**Comment by GrahamCampbell** [comment]
You're saying that "They have to implement HandlerInterface". That implies existing handlers are broken by definition.

**Comment by rtheunissen** [comment]
True, but the way they are used remains the same.

**Comment by rtheunissen** [comment]
Could change it to "they should implement"? :p

**Comment by rtheunissen** [comment]
Implemented handlers could be given a deprecation warning on a minor version update, or this can be included in the next major version.

**Comment by mtdowling** [comment]
I appreciate the effort here, but this is not the way that RingPHP was designed. See https://github.com/guzzle/RingPHP/issues/31

**Pull request closed without merging**